src/connection.rs

```diff
@@ -34,12 +73,22 @@ impl KafkaConnection {
     }
 
     pub fn new(host: &str, timeout_secs: i32) -> Result<KafkaConnection> {
-        let stream = try!(TcpStream::connect(host));
+        let stream = KafkaStream::Plain(try!(TcpStream::connect(host)));
         if timeout_secs > 0 {
             let t = Some(Duration::from_secs(timeout_secs as u64));
-            stream.set_read_timeout(t).expect("Set connection read-timeout");
-            stream.set_write_timeout(t).expect("Set connection write-timeout");
+            stream.get_ref().set_read_timeout(t).expect("Set connection read-timeout");
+            stream.get_ref().set_write_timeout(t).expect("Set connection write-timeout");
         }
         Ok(KafkaConnection{host: host.to_owned(), stream: stream})
     }
+
+    pub fn ssl(&mut self, config: Ssl) -> Result<()> {
+        let plain_stream = match self.stream {
+            KafkaStream::Plain(ref s) => try!(s.try_clone()),
+            KafkaStream::Ssl(_) => return Err(Error::AlreadySecure),
+        };
+        let stream = try!(SslStream::connect(config, plain_stream));
+        self.stream = KafkaStream::Ssl(stream);
+        Ok(())
+    }
```

Comment:
i would prefer, if we could avoid this public modifier method and move the ssl setup directly into the "`new` constructor function." What about passing a `Option<Ssl>`  as a third parameter there? This could avoid the error variant `Error::AlreadySecure` by definition.